pypiserver/core.py

```diff
@@ -97,279 +87,87 @@ def auth_by_htpasswd_file(htPsswdFile, username, password):
+        config.root = os.path.expanduser("~/packages")
+    roots = (
+        config.root if isinstance(config.root, (list, tuple)) else [config.root]
+    )
+    roots = [os.path.abspath(r) for r in roots]
+    for r in roots:
+        try:
+            os.listdir(r)
+        except OSError:
+            err = sys.exc_info()[1]
+            sys.exit(f"Error: while trying to list root({r}): {err}")
+    return roots
 
-    def fname_and_hash(self, hash_algo):
-        if not hasattr(self, "_fname_and_hash"):
-            if hash_algo:
-                self._fname_and_hash = (
-                    f"{self.relfn_unix}#{hash_algo}="
-                    f"{digest_file(self.fn, hash_algo)}"
-                )
-            else:
-                self._fname_and_hash = self.relfn_unix
-        return self._fname_and_hash
-
-
-def _listdir(root):
-    root = os.path.abspath(root)
-    for dirpath, dirnames, filenames in os.walk(root):
-        dirnames[:] = [x for x in dirnames if is_allowed_path(x)]
-        for x in filenames:
-            fn = os.path.join(root, dirpath, x)
-            if not is_allowed_path(x) or not os.path.isfile(fn):
-                continue
-            res = guess_pkgname_and_version(x)
-            if not res:
-                # #Seems the current file isn't a proper package
-                continue
-            pkgname, version = res
-            if pkgname:
-                yield PkgFile(
-                    pkgname=pkgname,
-                    version=version,
-                    fn=fn,
-                    root=root,
-                    relfn=fn[len(root) + 1 :],
-                )
-
-
-def read_lines(filename):
-    """
-    Read the contents of `filename`, stripping empty lines and '#'-comments.
-    Return a list of strings, containing the lines of the file.
-    """
-    lines = []
 
-    try:
-        with open(filename) as f:
-            lines = [
-                line
-                for line in (ln.strip() for ln in f.readlines())
-                if line and not line.startswith("#")
-            ]
-    except Exception:
-        log.error(
-            f'Failed to read package blacklist file "{filename}". '
-            "Aborting server startup, please fix this."
-        )
-        raise
+mimetypes.add_type("application/octet-stream", ".egg")
+mimetypes.add_type("application/octet-stream", ".whl")
+mimetypes.add_type("text/plain", ".asc")
 
-    return lines
 
+def get_bad_url_redirect_path(request, project):
+    """Get the path for a bad root url."""
+    uri = request.custom_fullpath
+    if uri.endswith("/"):
+        uri = uri[:-1]
+    uri = uri.rsplit("/", 1)[0]
+    project = quote(project)
+    uri += f"/simple/{project}/"
+    return uri
 
-def find_packages(pkgs, prefix=""):
-    prefix = normalize_pkgname(prefix)
-    for x in pkgs:
-        if prefix and x.pkgname_norm != prefix:
-            continue
-        yield x
 
+def with_digester(func: t.Callable[..., t.Iterable[PkgFile]]):
```

Comment:
to keep the decorator from overriding the underlying type, you can use a generic like:

```py
DigesterFunc = t.TypeVar("DigesterFunc", bound=t.Callable[..., t.Iterable[PkgFile]])

def with_digester(func: DigesterFunc) -> DigesterFunc:
  ...
```

you may need a `# type: ignore` where you're returning your wrapper function inside the decorator, but by telling mypy that you're returning a type of the same type that was put into the function, you allow it to infer that the resulting decorated function takes the same arguments as the original

Reply:
good idea! :)